### PR TITLE
refactor container top

### PR DESCRIPTION
**Pull request by ginglis13**
refactor the top command into pkg/cmd/container

CLI refactor main issue: #1680 

Signed-off-by: Gavin Inglis <giinglis@amazon.com>

### CODE REVIEW

File: pkg/cmd/container/top.go [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+	"github.com/containerd/containerd"
+	"github.com/containerd/nerdctl/pkg/api/types"
+	"github.com/containerd/nerdctl/pkg/clientutil"
+	"github.com/containerd/nerdctl/pkg/idutil/containerwalker"
+)
+
+// ContainerTopOKBody is from https://github.com/moby/moby/blob/v20.10.6/api/types/container/container_top.go
+//
+// ContainerTopOKBody OK response to ContainerTop operation
+type TopOKBody struct {
```
[the remark this answers is not in the document]
Comment (ginglis13) [reply pkg/cmd/container/top.go]: reverted back to `ContainerTopOKBody` in latest revision

File: cmd/nerdctl/top.go [hunk 2]
```diff
@@ -89,234 +81,8 @@ func topAction(cmd *cobra.Command, args []string) error {
...
-		return err
-	}
-
-	w := tabwriter.NewWriter(cmd.OutOrStdout(), 20, 1, 3, ' ', 0)
-	fmt.Fprintln(w, strings.Join(procList.Titles, "\t"))
-
-	for _, proc := range procList.Processes {
-		fmt.Fprintln(w, strings.Join(proc, "\t"))
-	}
+	return container.Top(cmd.Context(), args, cmd.OutOrStdout(), globalOptions)
```
[the remark this answers is not in the document]
Comment (ginglis13) [reply cmd/nerdctl/top.go]: changed the order of parameters in the latest revision 👍

File: pkg/cmd/container/top.go [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+	"github.com/containerd/containerd"
+	"github.com/containerd/nerdctl/pkg/api/types"
+	"github.com/containerd/nerdctl/pkg/clientutil"
+	"github.com/containerd/nerdctl/pkg/idutil/containerwalker"
+)
+
+// ContainerTopOKBody is from https://github.com/moby/moby/blob/v20.10.6/api/types/container/container_top.go
+//
+// ContainerTopOKBody OK response to ContainerTop operation
+type ContainerTopOKBody struct {
```
Comment (djdongjin) [inline pkg/cmd/container/top.go:40]: adding a `//nolint:revive` should avoid this CI error.

```golang
type ContainerTopOKBody struct { //nolint:revive
....
```

File: pkg/cmd/container/top.go [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+	"github.com/containerd/containerd"
+	"github.com/containerd/nerdctl/pkg/api/types"
+	"github.com/containerd/nerdctl/pkg/clientutil"
+	"github.com/containerd/nerdctl/pkg/idutil/containerwalker"
+)
+
+// ContainerTopOKBody is from https://github.com/moby/moby/blob/v20.10.6/api/types/container/container_top.go
+//
+// ContainerTopOKBody OK response to ContainerTop operation
+type ContainerTopOKBody struct { //nolint:revive
```
Comment (djdongjin) [inline pkg/cmd/container/top.go:40]: Can you remove the duplication in original place?

https://github.com/containerd/nerdctl/blob/ef78477d174b5c65ed72f4187eb6dddc7a3d4e06/cmd/nerdctl/top.go#L43

Comment (ginglis13) [reply pkg/cmd/container/top.go:40]: removed, thanks 😅

**Review by djdongjin — APPROVED** [review]
LGTM, thanks

File: cmd/nerdctl/top.go [hunk 5]
```diff
@@ -26,41 +26,17 @@
 package main
```
Comment (AkihiroSuda) [inline cmd/nerdctl/top.go:26]: > L17-L24

This notice should be moved too

File: pkg/cmd/container/top.go [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+	//
+	// Required: true
+	Processes [][]string `json:"Processes"`
+
+	// The ps column titles
+	// Required: true
+	Titles []string `json:"Titles"`
+}
+
+func Top(ctx context.Context, containers []string, opt types.GlobalCommandOptions, stdio io.Writer) error {
```
Comment (AkihiroSuda) [inline pkg/cmd/container/top.go:53]:
```suggestion
func Top(ctx context.Context, stdout io.Writer, containers []string, opt types.GlobalCommandOption) error {
```

File: cmd/nerdctl/top.go (hunk 5 above)
Comment (ginglis13) [reply cmd/nerdctl/top.go]: ack, moved

File: pkg/cmd/container/top.go (hunk 6 above)
Comment (ginglis13) [reply pkg/cmd/container/top.go]: created an options struct per other command refactors

File: pkg/api/types/container_types.go [hunk 7]
```diff
@@ -42,6 +42,13 @@ type ContainerStopOptions struct {
 	Timeout *time.Duration
 }
 
+// TopOptions specifies options for `nerdctl top`.
+type TopOptions struct {
```
Comment (djdongjin) [inline pkg/api/types/container_types.go:46]: Can you rename it to `ContainerTopOptions` to be consistent with other container options?

File: pkg/cmd/container/top.go [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+	Titles []string `json:"Titles"`
+}
+
+// Top performs the equivalent of running `top` inside of container(s)
+func Top(ctx context.Context, containers []string, opt types.TopOptions) error {
+	client, ctx, cancel, err := clientutil.NewClient(ctx, opt.GOptions.Namespace, opt.GOptions.Address)
+	if err != nil {
+		return err
+	}
+	defer cancel()
```
Comment (djdongjin) [inline pkg/cmd/container/top.go:68]: can you move client creation to cmd side? https://github.com/containerd/nerdctl/issues/1907

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Pull request merged**